File: binding/src/sql/libqsqldriverplugin.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsqldriverplugin.html
pub const qsqldriverplugin = struct {

    /// New constructs a new QSqlDriverPlugin object.
    pub fn New() QtC.QSqlDriverPlugin {
        return qtc.QSqlDriverPlugin_new();
    }


    /// New2 constructs a new QSqlDriverPlugin object.
    pub fn New2(parent: ?*anyopaque) QtC.QSqlDriverPlugin {
        return qtc.QSqlDriverPlugin_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDriverPlugin_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldriverplugin.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Create(self: ?*anyopaque, key: []const u8) QtC.QSqlDriver {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSqlDriverPlugin_Create(@ptrCast(self), key_str);
    }

    /// Allows for overriding the related default method
    pub fn OnCreate(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) QtC.QSqlDriver) void {
        qtc.QSqlDriverPlugin_OnCreate(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreate(self: ?*anyopaque, key: []const u8) QtC.QSqlDriver {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSqlDriverPlugin_QBaseCreate(@ptrCast(self), key_str);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDriverPlugin_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldriverplugin.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDriverPlugin_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldriverplugin.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSqlDriverPlugin_Delete(@ptrCast(self));
    }
};

File: binding/src/sql/libqsqldriverplugin.cpp
#include <QObject>
#include <QSqlDriver>
#include <QSqlDriverPlugin>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qsqldriverplugin.h>
#include "libqsqldriverplugin.h"
#include "libqsqldriverplugin.hxx"

QSqlDriverPlugin* QSqlDriverPlugin_new() {
	 return new VirtualQSqlDriverPlugin();
}

QSqlDriverPlugin* QSqlDriverPlugin_new2(QObject* parent) {
	 return new VirtualQSqlDriverPlugin(parent);
}

libqt_string QSqlDriverPlugin_Tr(const char* s) {
	QString _ret = QSqlDriverPlugin::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QSqlDriver* QSqlDriverPlugin_Create(QSqlDriverPlugin* self, const libqt_string key) {
	return self->create(QString::fromUtf8(key.data, key.len));
}

libqt_string QSqlDriverPlugin_Tr2(const char* s, const char* c) {
	QString _ret = QSqlDriverPlugin::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSqlDriverPlugin_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSqlDriverPlugin::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
QSqlDriver* QSqlDriverPlugin_QBaseCreate(QSqlDriverPlugin* self, const libqt_string key) {
	auto* vqsqldriverplugin = dynamic_cast<VirtualQSqlDriverPlugin*>(self);
	if (vqsqldriverplugin && vqsqldriverplugin->isVirtualQSqlDriverPlugin) {
vqsqldriverplugin->setQSqlDriverPlugin_Create_IsBase(true);
	return vqsqldriverplugin->create(QString::fromUtf8(key.data, key.len));
}
}

// Auxiliary method to allow providing re-implementation
void QSqlDriverPlugin_OnCreate(QSqlDriverPlugin* self, intptr_t slot) {
	auto* vqsqldriverplugin = dynamic_cast<VirtualQSqlDriverPlugin*>(self);
	if (vqsqldriverplugin && vqsqldriverplugin->isVirtualQSqlDriverPlugin) {
vqsqldriverplugin->setQSqlDriverPlugin_Create_Callback(reinterpret_cast<VirtualQSqlDriverPlugin::QSqlDriverPlugin_Create_Callback>(slot));
}
}

void QSqlDriverPlugin_Delete(QSqlDriverPlugin* self) {
    delete self;
}

